chapter05_recurrent-neural-networks/lstm-last-output.ipynb

```diff
@@ -0,0 +1,355 @@
+      "text/plain": [
+       "<matplotlib.figure.Figure at 0x119333e10>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "losses = []\n",
+    "for epoch in range(EPOCHS):\n",
+    "    for batch in np.random.choice(len(X), size=(int(len(X)/BATCH_SIZE), BATCH_SIZE)):\n",
+    "        with autograd.record():\n",
+    "            loss = sum(softmax_cross_entropy(net(nd.array(X[i]).reshape(shape=(-1, 1, 1))),\n",
+    "                                             nd.array([y[i]]).reshape((-1, 1)))\n",
+    "                       for i in batch)\n",
+    "        loss.backward()\n",
+    "        trainer.step(BATCH_SIZE)\n",
+    "        losses.append(loss.asscalar())\n",
+    "plt.plot(losses)\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Quick demo that it actually learned something,"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "array([[5480,    0],\n",
+       "       [ 971, 1115]])"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from sklearn.metrics import confusion_matrix\n",
+    "\n",
+    "y_pred = [nd.softmax(net(nd.array(x).reshape((-1, 1, 1))), axis=1).asnumpy().flatten().argmax() for x in X]\n",
+    "confusion_matrix(y, y_pred)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For the language example, lets try testing on a set of sentences the RNN hasn't seen."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[[417   0]\n",
+      " [224 165]]\n"
+     ]
+    }
+   ],
+   "source": [
+    "if EXAMPLE == 'lang':\n",
+    "    import requests, re\n",
+    "    novels = [\n",
+    "        (0, 'http://www.gutenberg.org/cache/epub/10822/pg10822.txt'), # Spanish\n",
+    "        (1, 'https://www.gutenberg.org/files/1342/1342-0.txt'), # English\n",
+    "    ]\n",
+    "    \n",
+    "    def sentences(novel, min_len=30, max_len=300):\n",
+    "        for s in re.sub(r'\\s+', ' ', requests.get(novel).text).split('.')[20:520]:\n",
+    "            if len(s) > min_len and len(s) < max_len:\n",
+    "                yield s\n",
+    "    \n",
+    "    # Run through all the novels and grab all the sentences in the books and grab all the sentences\n",
```

Comment:
What do you mean here?

Reply:
Just sentences out as series of characters